griptape/drivers/prompt/openai_chat_prompt_driver.py

```diff
@@ -7,58 +7,106 @@
             value will be bounded to the maximum possible as determined by the tokenizer.
         model: OpenAI model name. Uses `gpt-4` by default.
         organization: OpenAI organization.
         tokenizer: Custom `OpenAiTokenizer`.
-        user: OpenAI user. 	
+        user: OpenAI user.
     """
+
     api_type: str = field(default=openai.api_type, kw_only=True)
     api_version: Optional[str] = field(default=openai.api_version, kw_only=True)
     api_base: str = field(default=openai.api_base, kw_only=True)
     api_key: Optional[str] = field(default=Factory(lambda: os.environ.get("OPENAI_API_KEY")), kw_only=True)
     organization: Optional[str] = field(default=openai.organization, kw_only=True)
     model: str = field(kw_only=True)
     tokenizer: OpenAiTokenizer = field(
-        default=Factory(lambda self: OpenAiTokenizer(model=self.model), takes_self=True),
-        kw_only=True
+        default=Factory(lambda self: OpenAiTokenizer(model=self.model), takes_self=True), kw_only=True
     )
     user: str = field(default="", kw_only=True)
-    ignored_exception_types: Tuple[Type[Exception], ...] = field(default=Factory(lambda: (openai.InvalidRequestError)), kw_only=True)
+    ignored_exception_types: Tuple[Type[Exception], ...] = field(
+        default=Factory(lambda: openai.InvalidRequestError), kw_only=True
+    )
+    __ratelimit_request_limit: Optional[int] = field(init=False, default=None)
+    __ratelimit_requests_remaining: Optional[int] = field(init=False, default=None)
+    __ratelimit_requests_reset_at: Optional[datetime] = field(init=False, default=None)
+    __ratelimit_token_limit: Optional[int] = field(init=False, default=None)
+    __ratelimit_tokens_remaining: Optional[int] = field(init=False, default=None)
+    __ratelimit_tokens_reset_at: Optional[datetime] = field(init=False, default=None)
 
     def try_run(self, prompt_stack: PromptStack) -> TextArtifact:
+        # Define a hook to pull rate limit metadata from the OpenAI API response header.
+        openai.requestssession = requests.Session()
+        openai.requestssession.hooks = {"response": self._extract_ratelimit_metadata}
+
         result = openai.ChatCompletion.create(**self._base_params(prompt_stack))
 
         if len(result.choices) == 1:
-            return TextArtifact(
-                value=result.choices[0]["message"]["content"].strip()
-            )
+            return TextArtifact(value=result.choices[0]["message"]["content"].strip())
         else:
             raise Exception("Completion with more than one choice is not supported yet.")
 
     def token_count(self, prompt_stack: PromptStack) -> int:
-        return self.tokenizer.token_count(
-            self._prompt_stack_to_messages(prompt_stack)
-        )
+        return self.tokenizer.token_count(self._prompt_stack_to_messages(prompt_stack))
+
+    @property
+    def ratelimit_request_limit(self) -> Optional[int]:
```

Comment:
I think we can get rid of these properties now that the fields are not initialized up top. The purpose of this was to expose readonly public properties of private fields, right?

Reply:
Never-mind, we do want `@property`s if we want to expose these fields to the users. Though they add quite a bit to the class size, so _is_ this something we want to expose to the users? Especially since even we don't know how we're going to use them currently.

Reply:
Can't we just have it as `ratelimit_request_limit: Optional[int] = field(default=None)`?

Reply:
@vasinov That would allow it to be settable by the user, right? The approach here is to ensure it's a read-only property because, conceptually, it doesn't make sense to allow the user to write this value.

Reply:
Updated to protect the fields (`_fieldname`) and remove explicit getters.